packages/web-extension/src/walletManager/walletManagerWorker.ts

```diff
@@ -0,0 +1,128 @@
+import { BehaviorSubject, Observable, filter } from 'rxjs';
+import { Logger } from 'ts-log';
+import { ObservableWallet, storage } from '@cardano-sdk/wallet';
+import { Shutdown, isNotNil } from '@cardano-sdk/util';
+
+import { MessengerDependencies, MinimalRuntime, exposeApi } from '../messaging';
+import {
+  StoresFactory,
+  WalletFactory,
+  WalletManagerActivateProps,
+  WalletManagerApi,
+  WalletManagerProps
+} from './walletManager.types';
+import { consumeKeyAgent } from '../keyAgent';
+import { observableWalletProperties } from '../observableWallet';
+import { walletChannel } from './util';
+
+export interface WalletManagerDependencies {
+  walletFactory: WalletFactory;
+  storesFactory: StoresFactory;
+}
+
+/**
+ * Helper class for background scripts using wallet manager.
+ * Uses wallet and store factories to create wallets.
+ * Keeps track of created stores and reuses them when a wallet is reactivated.
+ */
+export class WalletManagerWorker implements WalletManagerApi {
+  #api$ = new BehaviorSubject<ObservableWallet | null>(null);
+  #hostSubscription: Shutdown;
+  #keyAgentSubscription: Shutdown;
+  #walletStores = new Map<string, storage.WalletStores>();
+
+  #walletFactory: WalletFactory;
+  #storesFactory: StoresFactory;
+  #logger: Logger;
+  #runtime: MinimalRuntime;
+
+  activeWallet$: Observable<ObservableWallet>;
+
+  constructor(
+    { walletName }: WalletManagerProps,
+    { walletFactory, storesFactory, logger, runtime }: MessengerDependencies & WalletManagerDependencies
+  ) {
+    this.activeWallet$ = this.#api$.asObservable().pipe(filter(isNotNil));
+    this.#walletFactory = walletFactory;
+    this.#storesFactory = storesFactory;
+    this.#logger = logger;
+    this.#runtime;
+    this.#hostSubscription = exposeApi(
+      {
+        api$: this.#api$.asObservable(),
+        baseChannel: walletChannel(walletName),
+        properties: observableWalletProperties
+      },
+      { logger, runtime }
+    );
+  }
+
+  async activate({ observableWalletName, cardanoServicesUrl }: WalletManagerActivateProps): Promise<void> {
+    if (await this.#isActive(observableWalletName)) {
+      return;
+    }
+    this.#deactivateWallet();
+
+    // Key agent is created by UI script on a per wallet unique channel
+    const keyAgent = consumeKeyAgent(
+      { walletName: observableWalletName },
+      { logger: this.#logger, runtime: this.#runtime }
+    );
+    this.#keyAgentSubscription = keyAgent;
+    const stores = this.#getStores(observableWalletName);
+
+    // Wallet factory is responsible for creating the wallet and the providers based on cardanoServicesUrl
```

Comment:
I agree `cardano-services` users can benefit from the convenience of just providing the API root URL, but this would be best done in the `cardano-services-client` package, with some abstract concept here to tie it together, as it's currently limiting the use of the manager to a single provider implementation. Think about the use case of local services, that may use Ogmios directly to submit transactions, and a HTTP server for the other providers.

Reply:
I suggest to change it to
```ts
// not destructuring
async activate(props: WalletManagerActivateProps): Promise<void> {
...
  await this.#walletFactory.create(props
```

Then WalletManagerActivateProps can become a union type to support other types of providers and this won't need to change.

Reply:
That implies all provider types are known to the SDK though, which isn't in the spirit of making this pluggable. It would also be good to not couple the convenience setup code to the `web-extension` context, so we can set up a client by passing a URL in other contexts, including Electron or Node.js.

Reply:
> That implies all provider implementations are known to the SDK though

Not necessarily, user can use his own walletFactory by design. However having that in mind, maybe we should change `WalletManagerActivateProps` to `{provider: {type: string; options: unknown;}}` instead of my suggestion [here](https://github.com/input-output-hk/cardano-js-sdk/pull/502#discussion_r1013948707)

> It would also be good to not couple the convenience setup code to the web-extension context...

Agree

Reply:
> user can use his own walletFactory 

Good point. Maybe the issue I have is just with the placement of the implementation

> maybe we should change...

Agree